### PR TITLE
Test container for postgres testing

### PR DESCRIPTION
**Pull request by MCatherine1994**

### CODE REVIEW

File: server/backend/testspg/router/test_router_application.py
Comment (MCatherine1994) [inline server/backend/testspg/router/test_router_application.py:11]: comment out this test, as it complains with the sqlite tests, but when run it locally without sqlite tests, it works fine

**Comment by basilv** [comment]
I took a look, but something more familiar with Python is better off reviewing than just me. The concept looks good, I'm not sure why we are keeping the SQLlite tests - is the plan to eventually eliminate those?

**Comment by ianliuwk1019** [comment]
New to this Testcontainers library, but first I notice the structure change and refactoring, that there is nested 'tests' folder under 'tests' folder. Is that necessary for the library to have nested '**tests/tests**' folder or just a temporary folder and have a plan to refactor again?

**Comment by ianliuwk1019** [comment]
Also quite some SonarCloud CodeSmells. I didn't look into it, might be more to do with naming consistency for Python.

**Comment by MCatherine1994** [comment]
> I took a look, but something more familiar with Python is better off reviewing than just me. The concept looks good, I'm not sure why we are keeping the SQLlite tests - is the plan to eventually eliminate those?

Hi Basil, yeah, I think the idea of this ticket is just to bring the concept to use test container for postgres testings. And just write some sample tests. And we'll open another ticket to write all tests and then remove our existing sqlite tests.

**Comment by MCatherine1994** [comment]
> New to this Testcontainers library, but first I notice the structure change and refactoring, that there is nested 'tests' folder under 'tests' folder. Is that necessary for the library to have nested '**tests/tests**' folder or just a temporary folder and have a plan to refactor again?

Hi Ian @ianliuwk1019, the only reason I changed the structure for the tests folder is because we said we don't want to remove all the sqlite tests for now, but the sqlite tests has some conflict with the new Postgres tests, so I have to put them in a subfolder. But in the future, once we write all the postgres tests, we can just remove the tests folder. The tests folder just holds the sqlite tests for now.

**Comment by ianliuwk1019** [comment]
> > New to this Testcontainers library, but first I notice the structure change and refactoring, that there is nested 'tests' folder under 'tests' folder. Is that necessary for the library to have nested '**tests/tests**' folder or just a temporary folder and have a plan to refactor again?
> 
> Hi Ian @ianliuwk1019, the only reason I changed the structure for the tests folder is because we said we don't want to remove all the sqlite tests for now, but the sqlite tests has some conflict with the new Postgres tests, so I have to put them in a subfolder to resolve the conflict. But in the future, once we write all the postgres tests, we can just remove the tests folder. The tests folder just holds the sqlite tests for now.



> > New to this Testcontainers library, but first I notice the structure change and refactoring, that there is nested 'tests' folder under 'tests' folder. Is that necessary for the library to have nested '**tests/tests**' folder or just a temporary folder and have a plan to refactor again?
> 
> Hi Ian @ianliuwk1019, the only reason I changed the structure for the tests folder is because we said we don't want to remove all the sqlite tests for now, but the sqlite tests has some conflict with the new Postgres tests, so I have to put them in a subfolder to resolve the conflict. But in the future, once we write all the postgres tests, we can just remove the tests folder. The tests folder just holds the sqlite tests for now.

I see. I do see you add 4 tests under new 'testpg' just for trying. I assume there will be not much change for all the current existing tests but to port them over and just have minor adjustment, right?
Curious, what was the issue you faced that you had to use nested 'test' folder to resolve them? I do not see a conflict (maybe there was something else...) between parent 'tests' folder and 'testspg' folder as they are in sibling level. Can't fix from within pytese config file?

**Comment by ianliuwk1019** [comment]
Hi Catherine,
Do you know why this 'testcontainers' is very slow?
We currently have about 60 existing Api tests.
In CI, tests running in Sqlite for all '60' tests only took - 20 seconds; while the '4' tests in new 'testspg' folder ran for 90 seconds, and the 4 are just simple tests. If in future, we run all 60 tests (and perhaps will be more and complex ones), according to this speed, the CI will take >30 minutes to finish just testing. It will greatly impact development.
From your past experience, is there a way to improve this? Or is it something that does not set up right?

**Comment by MCatherine1994** [comment]
> Hi Catherine,
> Do you know why this 'testcontainers' is very slow?
> We currently have about 60 existing Api tests.
> In CI, tests running in Sqlite for all '60' tests only took - 20 seconds; while the '4' tests in new 'testspg' folder ran for 90 seconds, and the 4 are just simple tests. If in future, we run all 60 tests (and perhaps will be more and complex ones), according to this speed, the CI will take >30 minutes to finish just testing. It will greatly impact development.
> From your past experience, is there a way to improve this? Or is it something that does not set up right?

Hi Ian, the reason it's a little bit slow it's because it has to setup and start a container for postgres and run flyway at the beginning. But the container will start once before all tests start, and stop after all test stop. So even we have more tests in the future, it won't increate more time cost because of the container.

**Comment by MCatherine1994** [comment]
> > > New to this Testcontainers library, but first I notice the structure change and refactoring, that there is nested 'tests' folder under 'tests' folder. Is that necessary for the library to have nested '**tests/tests**' folder or just a temporary folder and have a plan to refactor again?
> > 
> > 
> > Hi Ian @ianliuwk1019, the only reason I changed the structure for the tests folder is because we said we don't want to remove all the sqlite tests for now, but the sqlite tests has some conflict with the new Postgres tests, so I have to put them in a subfolder to resolve the conflict. But in the future, once we write all the postgres tests, we can just remove the tests folder. The tests folder just holds the sqlite tests for now.
> 
> > > New to this Testcontainers library, but first I notice the structure change and refactoring, that there is nested 'tests' folder under 'tests' folder. Is that necessary for the library to have nested '**tests/tests**' folder or just a temporary folder and have a plan to refactor again?
> > 
> > 
> > Hi Ian @ianliuwk1019, the only reason I changed the structure for the tests folder is because we said we don't want to remove all the sqlite tests for now, but the sqlite tests has some conflict with the new Postgres tests, so I have to put them in a subfolder to resolve the conflict. But in the future, once we write all the postgres tests, we can just remove the tests folder. The tests folder just holds the sqlite tests for now.
> 
> I see. I do see you add 4 tests under new 'testpg' just for trying. I assume there will be not much change for all the current existing tests but to port them over and just have minor adjustment, right? Curious, what was the issue you faced that you had to use nested 'test' folder to resolve them? I do not see a conflict (maybe there was something else...) between parent 'tests' folder and 'testspg' folder as they are in sibling level. Can't fix from within pytese config file?

Hi Ian @ianliuwk1019, yes, I just copied 4 existing tests and adjust them to test targeted on the postgres database. Since the purpose of this ticket is just to provide the concept, I didn't copy all the tests over, and that will need more time, and maybe create another ticket for it. 

The conflicts I got is that when I run the testspg test, the sqlite tests also try to get some sqlite connection, as shown in the screenshot. I'm not so sure why that happened, but I guess I can't have the origin conftest.py in the tests folder directly. I spent a lot of time try to solve this problem, but I still don't know how without changing the folder the structure. Since we'll get rid of the old tests anyway in the future, I'm not sure if we want to spend more time to figure out how to keep it work with the new tests.

Screen Shot 2023-03-02 at 12 12 39 PM

**Review by ianliuwk1019 — APPROVED** [review]
Approved this PR.
Maybe I am just being too cautious here, still feel the tests (4 simple tests in new 'Testcontainers') ran for 90 seconds seems a bit long (despite:  postgres db startup - assuming 10 ~15 seconds, and flyway migration - normally very fast, assuming 3 seconds), while sqllite for 60 tests ran for only 20 seconds.
Hopefully as you mentioned, newly migrated tests later will run fine within reasonable duration.

File: server/backend/tests/tests/conftest.py
Comment (MCatherine1994) [reply server/backend/tests/tests/conftest.py:58]: we have to comment out these, otherwise it complains "pytest_plugins" can only be at root level

Comment (ianliuwk1019) [reply server/backend/tests/tests/conftest.py:58]: Sorry, I meant, they can be removed, no need to commenting them out and leave them there.
Or do you still need to leave them as commented out code?

**Comment by MCatherine1994** [comment]
I didn't find a way to start the docker container just with postgres and flyway, and I asked this questions in the test-container GitHub repo, [testcontainers/testcontainers-python#315](https://github.com/testcontainers/testcontainers-python/issues/315). I'll merge this pull request for now, and if I heard anything from them, I'll update the ticket

**Comment by ianliuwk1019** [comment]
> I didn't find a way to start the docker container just with postgres and flyway, and I asked this questions in the test-container GitHub repo, [testcontainers/testcontainers-python#315](https://github.com/testcontainers/testcontainers-python/issues/315). I'll merge this pull request for now, and if I heard anything from them, I'll update the ticket

Thanks Catherine for looking into that.
Looks like the library is limited in this feature.
I guess then a different docker compose file (base compose file) for '**testcontainer**' might need for this concern on ignoring 'fam-api' setup. Would you mind adding this concern to the subsequent ticket's comment? But on the second thoughts, I am not that familiar with docker, but seems Docker compose file and do "extend" (from base docker compose file) and that probably is what we just need. Would be nice to see that in new ticket if you like to merge this first. Thanks.
@basilv

**Comment by basilv** [comment]
Trying to not launch the API container is just a nice-to-have speedup, I'm fine leaving as-is for now and focusing on getting rid of SQLlite.

**Comment by MCatherine1994** [comment]
> > I didn't find a way to start the docker container just with postgres and flyway, and I asked this questions in the test-container GitHub repo, [testcontainers/testcontainers-python#315](https://github.com/testcontainers/testcontainers-python/issues/315). I'll merge this pull request for now, and if I heard anything from them, I'll update the ticket
> 
> Thanks Catherine for looking into that. Looks like the library is limited in this feature. I guess then a different docker compose file (base compose file) for '**testcontainer**' might need for this concern on ignoring 'fam-api' setup. Would you mind adding this concern to the subsequent ticket's comment? But on the second thoughts, I am not that familiar with docker, but seems Docker compose file and do "extend" (from base docker compose file) and that probably is what we just need. Would be nice to see that in new ticket if you like to merge this first. Thanks. @basilv

I was originally having a separate docker compose file just to include postgres and flyway, but I feel there's no need to maintain two copies of the docker compose file just for testing. So I just used the one in the root directory. I'll leave it for now, and wait for their response, and add this task to the next ticker #469

**Pull request merged**
